plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/function/tablechanges/TableChangesSplitSource.java

```diff
@@ -0,0 +1,171 @@
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.ConnectorSplit;
+import io.trino.spi.connector.ConnectorSplitSource;
+import io.trino.spi.connector.SchemaTableName;
+import io.trino.spi.predicate.TupleDomain;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.deltalake.DeltaLakeErrorCode.DELTA_LAKE_INVALID_TABLE;
+import static io.trino.plugin.deltalake.function.tablechanges.DeltaLakeTableChangesFunction.FileSource.CDF_FILE;
+import static io.trino.plugin.deltalake.function.tablechanges.DeltaLakeTableChangesFunction.FileSource.DATA_FILE;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogUtil.getTransactionLogDir;
+import static io.trino.plugin.deltalake.transactionlog.checkpoint.TransactionLogTail.getEntriesFromJson;
+import static java.util.Objects.requireNonNull;
+
+public class TableChangesSplitSource
+        implements ConnectorSplitSource
+{
+    private final TrinoFileSystem fileSystem;
+    private final String tableLocation;
+    private final long endVersion;
+    private final AtomicBoolean isFinished;
+    private AtomicLong currentVersion;
+
+    public TableChangesSplitSource(
+            ConnectorSession session,
+            DeltaLakeMetastore metastore,
+            TrinoFileSystemFactory fileSystemFactory,
+            TableChangesFunctionTableHandle functionHandle)
+    {
+        requireNonNull(functionHandle, "functionHandle is null");
+        SchemaTableName schemaTableName = new SchemaTableName(functionHandle.getSchemaName(), functionHandle.getTableName());
+        this.tableLocation = requireNonNull(metastore, "metastore is null").getTableLocation(schemaTableName, session);
+        this.fileSystem = requireNonNull(fileSystemFactory, "fileSystemFactory is null").create(session);
+        currentVersion = new AtomicLong(functionHandle.getStartVersion());
+        endVersion = functionHandle.getEndVersion();
+        isFinished = new AtomicBoolean(false);
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+    {
+        long processedVersion = currentVersion.getAndIncrement();
+        if (processedVersion <= endVersion) {
+            Path transactionLogDir = getTransactionLogDir(new Path(tableLocation));
+            try {
+                Optional<List<DeltaLakeTransactionLogEntry>> entriesFromJson = getEntriesFromJson(processedVersion, transactionLogDir, fileSystem);
```

Comment:
What happens in case that the transaction log entry file is missing?
This may very well happen when `delta.logRetentionDuration` kicks in and the transaction log files before the last checkpoints get removed.

So if we request the changes from `100` to `200`, but we have transaction log entries starting from `180` (which is the last checkpoint of the table) should we get `180`-`200` range ?

Reply:
See https://github.com/trinodb/trino/pull/16192 which deals with this kind of situations for retrieving the `$history` of the table.

Reply:
> So if we request the changes from 100 to 200, but we have transaction log entries starting from 180 (which is the last checkpoint of the table) should we get 180-200 range ?

No, it should fail